gh-89004: add attribute filemode in zipinfo

Akasurde opened the pull request:
Add attribute zipfile.ZipInfo.filemode.

Fixes: #89004

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-89004 -->
* Issue: gh-89004
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--120999.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

Akasurde committed:
gh-89004: add attribute filemode in zipinfo

Add attribute zipfile.ZipInfo.filemode.

Fixes: #89004

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

eendebakpt commented on Misc/NEWS.d/next/Library/2024-06-25-07-49-16.gh-issue-89004.uJfz2B.rst:
```diff
@@ -0,0 +1 @@
+Add attribute :attr:`zipfile.ZipInfo.filemode`.
```
```suggestion
Add attribute :attr:`zipfile.ZipInfo.file_mode`.
```

eendebakpt commented on Lib/zipfile/__init__.py:
```diff
@@ -444,16 +446,20 @@ def _compresslevel(self):
                           compressor_names.get(self.compress_type,
                                                self.compress_type))
-        hi = self.external_attr >> 16
+        result.append(' filemode=%r' % self.file_mode)
```
```suggestion
        result.append(' file_mode=%r' % self.file_mode)
```
Would be more consistent with the name of the attribute. But that does change the behaviour of the repr, so I am not sure on this one.

eendebakpt commented on Lib/zipfile/__init__.py:
```diff
@@ -444,16 +446,20 @@ def _compresslevel(self):
         self.compress_level = value
 
+    @property
+    def _file_mode(self):
```
Why is the property called `_file_mode` and not `file_mode`?

kumaraditya303 committed:
gh-87744: fix waitpid race while calling send_signal in asyncio (#121126)

asyncio earlier relied on subprocess module to send signals to the process, this has some drawbacks one being that subprocess module unnecessarily calls waitpid on child processes and hence it races with asyncio implementation which internally uses child watchers. To mitigate this, now asyncio sends signals directly to the process without going through the subprocess on non windows systems. On Windows it fallbacks to subprocess module handling but on windows there are no child watchers so this issue doesn't exists altogether.

eendebakpt commented on Lib/zipfile/__init__.py:
```diff
@@ -430,6 +431,7 @@ def __init__(self, filename="NoName", date_time=(1980,1,1,0,0,0)):
         self.external_attr = 0          # External file attributes
         self.compress_size = 0          # Size of the compressed file
         self.file_size = 0              # Size of the uncompressed file
+        self.file_mode = b""            # External attributes of the uncompressed file
```
Why is this line needed?

eendebakpt commented on Lib/zipfile/__init__.py:
```diff
@@ -444,16 +446,20 @@ def _compresslevel(self):
     def _compresslevel(self, value):
         self.compress_level = value
 
+    @property
```
The new property needs unit testing for different modes

eendebakpt reviewed (commented):
@Akasurde Two more comments, but the overall approach looks good.

Please do not use force push on the branch, it makes it hard for other reviewers the follow the history.

erlend-aasland committed:
gh-113565: Improve and harden detection of curses dependencies (#119816)

1. Use pkg-config to check for ncursesw/panelw. If that fails, use 
   pkg-config to check for ncurses/panel.
2. Regardless of pkg-config output, search for curses/panel headers, so
   we're sure we have all defines in pyconfig.h.
3. Regardless of pkg-config output, check if libncurses or libncursesw
   contains the 'initscr' symbol; if it does _and_ pkg-config failed
   earlier, add the resulting -llib linker option to CURSES_LIBS.
   Ditto for 'update_panels' and PANEL_LIBS.
4. Wrap the rest of the checks with WITH_SAVE_ENV and make sure we're 
   using updated LIBS and CPPFLAGS for those.

Add the PY_CHECK_CURSES convenience macro.

vstinner committed:
gh-121188: Sanitize invalid XML characters in regrtest (#121195)

When creating the JUnit XML file, regrtest now escapes characters
which are invalid in XML, such as the chr(27) control character used
in ANSI escape sequences.

vstinner committed:
gh-121199: Use _Py__has_attribute() in timemodule.c (#121203)

Use the _Py__has_attribute() macro in timemodule.c and
bootstrap_hash.c to fix a build error on old GCC versions (GCC 4.8.5
on s390x).

skirpichev committed:
gh-61103: Support double complex (_Complex) type in ctypes (#120894)

Example:

```pycon
>>> import ctypes
>>> ctypes.__STDC_IEC_559_COMPLEX__
1
>>> libm = ctypes.CDLL('libm.so.6')
>>> libm.clog.argtypes = [ctypes.c_double_complex]
>>> libm.clog.restype = ctypes.c_double_complex
>>> libm.clog(1+1j)
(0.34657359027997264+0.7853981633974483j)
```

Co-authored-by: Nice Zombies <nineteendo19d0@gmail.com>
Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>
Co-authored-by: Victor Stinner <vstinner@python.org>

vstinner committed:
gh-121084: Call _abc_registry_clear() when checking refleaks (#121191)

dash_R_cleanup() now calls _abc_registry_clear() before calling again
register().

vstinner committed:
gh-121200: Log pwd entry in test_expanduser_pwd2() (#121207)

Use subTest() to log the pwd entry in test_expanduser_pwd2() of
test_posixpath to help debugging.

dependabot[bot] committed:
build(deps-dev): bump types-psutil from 5.9.5.20240516 to 6.0.0.20240621 in /Tools (#121215)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps-dev): bump mypy from 1.10.0 to 1.10.1 in /Tools (#121216)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps-dev): bump types-setuptools from 70.0.0.20240524 to 70.1.0.20240627 in /Tools (#121217)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

WillChilds-Klein committed:
gh-117784: Only reference PHA functions ifndef SSL_VERIFY_POST_HANDSHAKE (GH-117785)

With this change, builds with OpenSSL forks that don't have this functionalty
(like AWS-LC or BoringSSL) will require less patching.

Eclips4 committed:
gh-121220: Mark test_threaded_weak_value_dict_copy() as CPU-heavy (#121221)

Mark test_threaded_weak_value_dict_copy() and
test_threaded_weak_key_dict_copy() of test_weakref as
CPU-heavy tests

aisk committed:
gh-117657: Use critical section to make _socket.socket.close thread safe (GH-120490)

vstinner committed:
gh-121200: Fix test_expanduser_pwd2() of test_posixpath (#121228)

Call getpwnam() to get pw_dir, since it can be different than
getpwall() pw_dir.

vstinner committed:
gh-120743: Soft deprecate os.popen() function (#120744)

Soft deprecate os.popen() and os.spawn*() functions.

yangdanny97 committed:
gh-114104: clarify asynchronous comprehension docs to match runtime behavior (#121175)

ericsnowcurrently committed:
gh-121110: Temporarily Skip test_basic_multiple_interpreters_reset_each (gh-121236)

This will allow Py_TRACE_REFS builds to pass the test suite, until the underlying issue can be resolved.

aisk committed:
gh-117657: Fix data races reported by TSAN in some set methods (#120914)

Refactor the fast Unicode hash check into `_PyObject_HashFast` and use relaxed
atomic loads in the free-threaded build.

After this change, the TSAN doesn't report data races for this method.

brandtbucher committed:
GH-116017: Get rid of _COLD_EXITs (GH-120960)

sobolevn committed:
gh-121196: Document `dict.fromkeys` params as pos-only (#121197)

diegorusso committed:
GH-119726: Use LDR for AArch64 trampolines (GH-121001)

rffontenelle committed:
Fix phrasing in paragraphs with leading "similar" (#121135)

vsajip committed:
gh-121035: Improve logging flow diagram for dark/light modes. (GH-121254)

vstinner committed:
Move get_signal_name() to test.support (#121251)

* Move get_signal_name() from test.libregrtest to test.support.
* Use get_signal_name() in support.script_helper.
* support.script_helper now decodes stdout and stderr from UTF-8,
  instead of ASCII, if a command failed.

skirpichev committed:
gh-121245: Amend d611c4c8e9 (correct import) (#121255)

Co-authored-by: Miro Hrončok <miro@hroncok.cz>

picnixz committed:
gh-121210: handle nodes with missing attributes/fields in `ast.compare` (#121211)

picnixz committed:
gh-121165: protect macro expansion of `ADJUST_INDICES` with do-while(0) (#121166)

iritkatriel committed:
gh-121272: move __future__ import validation from compiler to symtable (#121273)

colesbury committed:
gh-117139: Add _PyTuple_FromStackRefSteal and use it (#121244)

Avoids the extra conversion from stack refs to PyObjects.

pablogsal committed:
gh-115773: Add sizes to debug offset structure (#120112)

vsajip committed:
gh-121035: Further improve logging flow diagram with respect to dark/light modes. (GH-121265)

Akasurde committed:
CI fixes

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>
